src/module/service.cpp

```diff
@@ -210,6 +215,11 @@ void ModuleService_::start() {
     std::string address = "unix://" + module->addr;
     Server::add_listening_port(address);
 
+    // Call exit(0) on SIGTERM or SIGINT so module can cleanly shutdown with its
+    // destructor.
+    signal(SIGINT, signal_handler);
```

Comment:
Similar thoughts as above: library code shouldn't be making decisions about the disposition or handling of signals. That's the responsibility of the program.

Reply:
While our handling of `SIGINT` and `SIGTERM` in CPP modules is more explicit here, Go and Python do the same sort of "swallowing" of the signals and execution of a graceful shutdown within their libraries (the graceful handling of those signals is [a requirement](https://github.com/viamrobotics/rdk/blob/main/module/doc.go#L65) of module design). The example modules in those languages do not do any sort of signal handling and instead rely on the underlying library. I do agree that it's unorthodox to handle these signals within the library, but it also seems like a potentially unnecessary burden to hand over to users. It seems relatively easy for a user to incorrectly handle these signals and shoot themselves in the foot. What are your thoughts @stuqdog ?

Reply:
I'm inclined to keep the code within the library. While I agree that it's a strange pattern, I think I favor consistency with the other SDKs - and with the `module` spec - in this case. Since the `module` spec specifically says that what is built, I think there's potential for additional confusion beyond what @benjirewis highlighted above if we fail to handle shutdown. Happy to keep the conversation going if there's a strong contrary preference, however.

Reply:
- Regarding the requirement that modules handle signals gracefully, I think that the requirement is good from the perspective of a module as an executable that gets run. But that does*not mean that the SDK used to produce modules must itself automatically handle the signals. As long as the author of each module ensures that signals are properly handled, the requirement is still met, though with some unfortunate complexity pushed to the module author.

- The C++ SDK is not in a position to fully know what should be done on receipt of a signal because it cannot know what requirements *other* pieces of software in use by the module may have. For instance, if you were using the C++ SDK to produce a modular resource that supported the operation of an arm, that arm may come with its own SDK with its own shutdown requirements. Only the author of the module is in a position to coordinate the graceful termination of all of the components in use, of which the Viam C++ SDK is only one.

- It isn't actually legal per the standard to go invoking destructors in a signal handler. In fact, it isn't legal to do more or less anything in a signal handler except call `_Exit` (no dtors run) or `quick_exit` (ditto), or `std::abort` (same). About the only thing you are allowed to do is toggle an atomic flag, and the rules get even crazier in C++17. Even if C++ weren't in play you are still more or less restricted to calling the functions in the "async signal safe" set, which is very limited (https://man7.org/linux/man-pages/man7/signal-safety.7.html). A better way to go about all of this is to use https://man7.org/linux/man-pages/man2/sigwaitinfo.2.html instead of `signal`. But again, that's on the module author, not the SDK.

What you could do, and might make sense, would be to offer a free function in the SDK somewhere that sets up a minimal and correct signal handling environment under the assumption that there are no other shutdown requirements, and then have the example module call that, since in the examples there are no other requirements than tearing down the SDK correctly. That would make it easy for authors of simple modules to obtain the required signal handling behavior with low effort, but not put the SDK in the position of taking on a burden it can't actually handle in the more complex cases.

Reply:
Thanks for all the info @acmorrow ; it seems like it's neither feasible nor optimal to handle signals from within the SDK. I've moved the signal handling to the example module. Let me know what you think.